### PR TITLE
Implement Basic Gluon

### PR DESCRIPTION
**Pull request by toddsundsted**
Enough Gluon to get through the first few chapters of The Straight Dope, including basic Perceptrons and CNNs. Also includes Python Gluon-compatible support for hybridization and export/import of models and parameters.

@mrkn mxnet.rb is an excellent library. Hopefully you find this a useful addition. I would appreciate feedback and comments, in any case.

### CODE REVIEW

**Comment by mrkn** [comment]
@toddsundsted  I very appreciate this pull-request. Thank you for your work.

I've already been working on implementing Gluon APIs in the local branch based on [this temporary working branch](https://github.com/mrkn/mxnet.rb/tree/taiwan2018/lib/mxnet), so let me take several days to check how to integrate both of them.

Thank you very much again.

**Comment by toddsundsted** [comment]
@mrkn awesome! let me know if I can help in any way.

File: ext/mxnet/cached_op.c [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  cargs.num_inputs = (mx_uint)n;
+  cargs.inputs = ALLOC_N(NDArrayHandle, n);
+
+  for (i = 0; i < n; ++i) {
+    cargs.inputs[i] = mxnet_ndarray_get_handle(RARRAY_AREF(args, i));
+  }
+
+  cargs.cached_op_handle = (CachedOpHandle)DATA_PTR(obj);
+
+  result = rb_protect((VALUE (*)(VALUE))call_invoke_cached_op, (VALUE)&cargs, &state);
```
Comment (mrkn) [inline ext/mxnet/cached_op.c]: @toddsundsted Would you explain why calling MXInvokeCachedOpEx is necessary to be enclosed in rb_protect?

Comment (mrkn) [reply ext/mxnet/cached_op.c]: Ah, I found the reason.  That is to free memories allocated above.

Comment (toddsundsted) [reply ext/mxnet/cached_op.c]: @mrkn thanks for pointing this out. looking at the code. the strategy is to use an allocated ruby array like `gc_guard` to hold ruby values for keys and values, and to rely on the garbage collector to reclaim the memory regardless of whether there was an error or not?

Comment (mrkn) [reply ext/mxnet/cached_op.c]: @toddsundsted Yes, that's right.

Comment (mrkn) [reply ext/mxnet/cached_op.c]: @toddsundsted I think #40 is a good exercise for using `rb_str_tmp_new` and a gc_guard array.

Comment (toddsundsted) [reply ext/mxnet/cached_op.c]: @mrkn alright! i'll fix it!

**Comment by mrkn** [comment]
This PR was resolved by #36

**Pull request closed without merging**